blog/2025-11-11-scaling-kafka-with-automq-and-tigris/index.mdx

```diff
@@ -0,0 +1,203 @@
+guarantees ordering within a partition and durable persistence on disk.
+
+![](./automq-figure-2.webp)
+
+### Durability & replication
+
+Kafka ensures durability by maintaining at least 3 replicas per partition. Each
+partition has a leader and multiple followers forming an In-Sync Replicas (ISR)
+that must be nearly caught up with the leader's writes.
+
+![](./automq-figure-3.webp)
+
+The replication flow is as follows: Producer writes to the leader. The leader
+writes to disk via page cache. Followers then fetch the new data from the leader
+that they persist and report back to the leader.
+
+The downside with this approach is that every partition write becomes a fan-out
+stream to multiple brokers. At high throughput, the internal replication traffic
+dominates network usage, which forces operators to scale to larger EC2 instances
+and/or higher-bandwidth connections.
+
+### How AutoMQ and Tigris solve durability and replication
+
+AutoMQ offloads the durability layer to Tigris as it inherently provides
+multiple replicas, offering extremely high data durability. Consequently, AutoMQ
+offers higher throughput by eliminating the need for data redundancy replication
+overhead.
+
+![](./automq-figure-4.webp)
+
+A producer writes to any AutoMQ broker. The broker streams log segments directly
+into Tigris. Tigris stores each segment as an immutable object and replicates it
+across regions. Through this design:
+
+- All brokers can access the same data without replication
+- There is no follower lag or ISR management
+- Write throughput increases because internal fan-out disappears
+
+AutoMQ retains Kafka's protocol behavior while removing the operational weight
+of traditional replication.
+
+Additionally, Tigris distributes the data globally, ensuring regional latencies
+for Producers and Consumers.
+
+![](./automq-figure-5.webp)
+
+The diagram illustrates how AutoMQ operates when backed by Tigris object
+storage. A producer sends records to any AutoMQ broker, which remains stateless
+and handles only protocol-level logic. Instead of writing log segments to local
+disk or replicating them across brokers, the broker streams each segment
+directly into Tigris. Tigris stores these segments as immutable objects per
+partition (P1, P2, P3) and replicates them across multiple regions. Because
+replication happens entirely within the storage layer, all brokers in any region
+see the same partition data without ISR management, follower lag, or
+inter-broker traffic.
+
+## Partition and cluster rebalancing
+
+### The challenges with rebalancing Kafka
+
+Kafka stores data locally on each broker. When the cluster scales in or out,
+Kafka must:
+
+- Copy partition data to new brokers
+- Rebuild replicas
+- Re-distribute leadership
+
+This involves heavy disk I/O, large data movement, and long rebalancing
+workflows. Redistributing data can take hours or days, disrupting cluster
+performance.
```

Comment:
This feels insufficiently detailed. If I have moderately large SSDs, highly parallel CPUs, more than enough RAM, and multi-gigabit networking, why do I care that the cluster is balancing its humors?